x-pack/filebeat/input/cometd/input.go

```diff
@@ -0,0 +1,204 @@
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package cometd
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/v7/filebeat/channel"
+	"github.com/elastic/beats/v7/filebeat/input"
+	"github.com/elastic/beats/v7/libbeat/beat"
+	"github.com/elastic/beats/v7/libbeat/common"
+	"github.com/elastic/beats/v7/libbeat/common/atomic"
+	"github.com/elastic/beats/v7/libbeat/common/transport/httpcommon"
+	"github.com/elastic/beats/v7/libbeat/logp"
+
+	bay "github.com/elastic/bayeux"
+)
+
+const (
+	inputName = "cometd"
+)
+
+// Run starts the input worker then returns. Only the first invocation
+// will ever start the worker.
+func (in *cometdInput) Run() {
+	in.workerOnce.Do(func() {
+		in.workerWg.Add(1)
+		go func() {
+			in.log.Info("Input worker has started.")
+			defer in.log.Info("Input worker has stopped.")
+			defer in.workerWg.Done()
+			defer in.workerCancel()
+			if err := in.run(); err != nil {
+				in.log.Error(err)
+				return
+			}
+		}()
+	})
+}
+
+func (in *cometdInput) run() error {
+	b := bay.Bayeux{}
+	creds := bay.GetSalesforceCredentials()
+	in.out = b.Channel(in.out, "-1", creds, in.config.ChannelName)
+
+	var event Event
+	for e := range in.out {
+		if !e.Successful {
+			// To handle the last response where the object received was empty
+			if e.Data.Payload == nil {
+				return nil
+			}
+
+			// Convert json.RawMessage response to []byte
+			msg, err := json.Marshal(e.Data.Payload)
+			if err != nil {
+				return fmt.Errorf("JSON error: %v", err)
+			}
+
+			// Extract event IDs from json.RawMessage
+			err = json.Unmarshal(e.Data.Payload, &event)
+			if err != nil {
+				return fmt.Errorf("error while parsing JSON: %v", err)
+			}
+			if ok := in.outlet.OnEvent(makeEvent(event.EventId, string(msg))); !ok {
+				in.log.Debug("OnEvent returned false. Stopping input worker.")
+				close(in.out)
+				return fmt.Errorf("error ingesting data to elasticsearch")
+			}
+		}
+	}
+	return nil
+}
+
+func init() {
+	err := input.Register(inputName, NewInput)
+	if err != nil {
+		panic(errors.Wrapf(err, "failed to register %v input", inputName))
+	}
+}
+
+// NewInput creates a new CometD input that consumes events from
+// a topic subscription.
+func NewInput(
+	cfg *common.Config,
+	connector channel.Connector,
+	inputContext input.Context,
+) (inp input.Input, err error) {
+	// Extract and validate the input's configuration.
+	conf := defaultConfig()
+	if err = cfg.Unpack(&conf); err != nil {
+		return nil, err
+	}
+
+	os.Setenv("SALESFORCE_CONSUMER_KEY", conf.Auth.OAuth2.ClientID)
+	os.Setenv("SALESFORCE_CONSUMER_SECRET", conf.Auth.OAuth2.ClientSecret)
+	os.Setenv("SALESFORCE_USER", conf.Auth.OAuth2.User)
+	os.Setenv("SALESFORCE_PASSWORD", conf.Auth.OAuth2.Password)
+	os.Setenv("SALESFORCE_TOKEN_URL", conf.Auth.OAuth2.TokenURL)
```

Comment:
Why do we need to set those?

Reply:
We are setting them because of the way it has been implemented [here](https://github.com/elastic/bayeux/blob/3678496b62f26e10595b48068714c2b7eb2334f6/bayeux.go#L241) in Bayeux.

Reply:
As we own the fork of `zph/bayeux`, we can adjust the implementation there. This is my current recommendation as setting all these envs is a spaghetti code.

Reply:
Raised a PR https://github.com/elastic/bayeux/pull/2/ to accommodate the same in Bayeux.